File: backend/app/intent/handlers/delete_form_handler.py
# DeleteFormHandler - 删除表单意图处理器
# 对应 chat.py 第918-972行

import asyncio
import time
from typing import AsyncGenerator

from ..base import BaseIntentHandler, IntentContext
from ..utils import thinking, sse, intent_event, done_event
from ...services.admin_service import AdminService


class DeleteFormHandler(BaseIntentHandler):
    """删除表单意图 —— 自动备份 + 删除"""

    intent_type = "delete_form"

    async def handle(self, ctx: IntentContext) -> AsyncGenerator[str, None]:
        target_code = ctx.intent_data.get("formCode", "") or ctx.intent_data.get("detectedFormCode", "")
        target_name = ctx.intent_data.get("formName", "")
        if not target_name and target_code and target_code in ctx.ontologies:
            target_name = ctx.ontologies[target_code].get("formName", target_code)

        # ── Phase 1：确认删除目标 ────────────────────────────────────
        yield thinking(f"🗑️ 确认删除表单「{target_name or target_code}」（自动备份）", result={
            "formCode": target_code,
            "formName": target_name,
            "autoBackup": True
        })

        # 执行删除
        loop = asyncio.get_event_loop()
        delete_result = await loop.run_in_executor(
            None,
            lambda: AdminService.delete_ontology(target_code, auto_backup=True)
        )

        ctx.stream_stats.total_elapsed = time.time() - ctx.start_time

        if delete_result.get("success"):
            backup = delete_result.get("backup", {})
            backup_id = backup.get("id", "") if backup else ""

            # ── Phase 2：删除结果 ────────────────────────────────────
            yield thinking(f"✅ 已删除表单「{target_name or target_code}」", result={
                "success": True,
                "backupVersionId": backup_id,
                "message": delete_result.get("message", "")
            })

            yield sse({"type": "stats", "content": ctx.stream_stats.to_dict()})

            yield intent_event("delete_form", "delete", {
                "formCode": target_code,
                "formName": target_name,
                "backupVersionId": backup_id,
                "message": delete_result.get("message", "")
            }, is_form=False)

            confirm_text = f"已删除表单「{target_name or target_code}」。如需恢复，可以在版本历史中回退。"
            yield sse({"type": "text_start"})
            for i in range(0, len(confirm_text), 3):
                yield sse({"type": "text", "content": confirm_text[i:i+3]})
                await asyncio.sleep(0.02)
            yield sse({"type": "text_end"})

            yield done_event("delete_form", is_form=False, intent_data=ctx.intent_data)
        else:
            error_msg = delete_result.get("message", "删除失败")

            # ── Phase 2：错误结果（合并到同一结构） ───────────────────
            yield thinking(f"❌ 删除失败: {error_msg}", result={
                "success": False,
                "error": error_msg
            })

            ctx.stream_stats.error = error_msg
            yield sse({"type": "stats", "content": ctx.stream_stats.to_dict()})
            yield sse({"type": "error", "content": error_msg})
            yield done_event("delete_form", is_form=False, intent_data=ctx.intent_data)

File: backend-app/src/main/java/com/sitech/prodai/intent/handlers/DeleteFormHandler.java
package com.sitech.prodai.intent.handlers;

import com.sitech.prodai.intent.BaseIntentHandler;
import com.sitech.prodai.intent.IntentContext;
import com.sitech.prodai.intent.SseUtils;
import com.sitech.prodai.intent.StreamStats;
import com.sitech.prodai.service.OntologyService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Schedulers;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 删除表单意图处理器 —— 对齐 Python {@code app/intent/handlers/delete_form_handler.py::DeleteFormHandler}。
 *
 * <p>自动备份 + 删除。Python 端调用 AdminService.delete_ontology，
 * Java 端使用 {@link OntologyService}（文件数据源模式下写操作返回"不支持"）。
 */
@Component
public class DeleteFormHandler implements BaseIntentHandler {

    private static final Logger log = LoggerFactory.getLogger(DeleteFormHandler.class);

    private final OntologyService ontologyService;

    public DeleteFormHandler(OntologyService ontologyService) {
        this.ontologyService = ontologyService;
    }

    @Override
    public String getIntentType() {
        return "delete_form";
    }

    @Override
    public Flux<Map<String, Object>> handle(IntentContext ctx) {
        Map<String, Object> intentData = ctx.getIntentData();
        String targetCode = firstNonBlank(str(intentData.get("formCode")), str(intentData.get("detectedFormCode")));
        String targetName = str(intentData.get("formName"));
        if (targetName.isEmpty() && targetCode != null && ctx.getOntologies().containsKey(targetCode)) {
            targetName = str(ctx.getOntologies().get(targetCode).get("formName"));
        }

        final String code = targetCode != null ? targetCode : "";
        final String name = targetName;

        // Phase 1：识别
        Map<String, Object> identifyResult = new LinkedHashMap<>();
        identifyResult.put("formCode", code);
        identifyResult.put("formName", name);
        identifyResult.put("autoBackup", true);

        return Flux.concat(
                Flux.just(SseUtils.thinking("🗑\uFE0F 确认删除表单「" + (name.isEmpty() ? code : name) + "」（自动备份）", identifyResult)),

                // Phase 2：执行（在 boundedElastic 线程池中执行阻塞调用）
                reactor.core.publisher.Mono.fromCallable(() -> ontologyService.deleteOntology(code))
                        .subscribeOn(Schedulers.boundedElastic())
                        .flatMapMany(deleteResult -> {
                            List<Map<String, Object>> events = new ArrayList<>();
                            StreamStats stats = ctx.getStreamStats();
                            if (stats != null) {
                                stats.setTotalElapsed((System.currentTimeMillis() - ctx.getStartTime()) / 1000.0);
                            }

                            if (Boolean.TRUE.equals(deleteResult.get("success"))) {
                                String backupId = "";
                                Object backup = deleteResult.get("backup");
                                if (backup instanceof Map<?, ?> backupMap) {
                                    backupId = str(backupMap.get("id"));
                                }

                                Map<String, Object> successResult = new LinkedHashMap<>();
                                successResult.put("success", true);
                                successResult.put("backupVersionId", backupId);
                                successResult.put("message", str(deleteResult.get("message")));

                                events.add(SseUtils.thinking("✅ 已删除表单「" + (name.isEmpty() ? code : name) + "」", successResult));
                                if (stats != null) {
                                    events.add(SseUtils.stats(stats));
                                }

                                Map<String, Object> intentEventData = new LinkedHashMap<>();
                                intentEventData.put("formCode", code);
                                intentEventData.put("formName", name);
                                intentEventData.put("backupVersionId", backupId);
                                intentEventData.put("message", str(deleteResult.get("message")));
                                events.add(SseUtils.intentEvent("delete_form", "delete", intentEventData, false));

                                // 流式输出确认文本
                                String confirmText = "已删除表单「" + (name.isEmpty() ? code : name) + "」。如需恢复，可以在版本历史中回退。";
                                events.add(SseUtils.textStart());
                                for (int i = 0; i < confirmText.length(); i += 3) {
                                    events.add(SseUtils.text(confirmText.substring(i, Math.min(i + 3, confirmText.length()))));
                                }
                                events.add(SseUtils.textEnd());
                                events.add(SseUtils.doneEvent("delete_form", false, ctx.getIntentData()));
                            } else {
                                String errorMsg = str(deleteResult.get("message"));
                                if (errorMsg.isEmpty()) {
                                    errorMsg = "删除失败";
                                }
                                Map<String, Object> failResult = new LinkedHashMap<>();
                                failResult.put("success", false);
                                failResult.put("error", errorMsg);
                                events.add(SseUtils.thinking("❌ 删除失败: " + errorMsg, failResult));
                                if (stats != null) {
                                    stats.setError(errorMsg);
                                    events.add(SseUtils.stats(stats));
                                }
                                events.add(SseUtils.error(errorMsg));
                                events.add(SseUtils.doneEvent("delete_form", false, ctx.getIntentData()));
                            }
                            return Flux.fromIterable(events);
                        })
        );
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value);
    }

    private String firstNonBlank(String... values) {
        if (values == null) {
            return null;
        }
        for (String v : values) {
            if (v != null && !v.isBlank()) {
                return v;
            }
        }
        return null;
    }
}
